### PR TITLE
DRILL-6591: Show Exception for failed queries submitted in WebUI

### PR DESCRIPTION
**Pull request by kkhatua**
When query fails on Web UI result page no error is shown, only "No result found." 
This was because DRILL-6477 (PR #1309) switched to `WebUserConnection.await(long timeoutInMillis)` . Unlike the original `WebUserConnection.await()`, this method did not throw any UserException generated by a query failure.
The fix was to add a new WebUser-only method - `WebUserConnection.timedWait(long timeoutInMillis)` 
This ensures that other callers to the `WebUserConnection.await(timeoutInMillis)` are unaffected by the expected behaviour.

This is what the UI returns after this patch:
_When running query which hits memory limits (DRILL-6477), we get this:_
```
org.apache.drill.common.exceptions.UserException: RESOURCE ERROR: There is not enough heap memory to run this query using the web interface. Please try a query with fewer columns or with a filter or limit condition to limit the data returned. You can also try an ODBC/JDBC client. [Error Id: 702884d0-4e90-46c0-a861-71b7d384084b on kk127.qa.lab:31010]
```

_When querying a non-existent schema:_
```
org.apache.drill.common.exceptions.UserRemoteException: VALIDATION ERROR: Schema [[sys1]] is not valid with respect to either root schema or current default schema. Current default schema: No default schema selected [Error Id: 9fd6041c-b4f1-4f63-912e-d8a44ed318b0 on kk127.qa.lab:31010]
```

_When running an incorrect SQL:_
```
Non-query expression encountered in illegal context (Non-query expression encountered in illegal context) org.apache.drill.common.exceptions.UserException: PARSE ERROR: Non-query expression encountered in illegal context [Error Id: 22238845-2514-4bb0-af8a-be5b6c922a63 on kk127.qa.lab:31010]
```

### CODE REVIEW

**Comment by kkhatua** [comment]
@arina-ielchiieva / @sohami  could you please review this?
(Thanks for finding the bug and giving pointers to the fix)

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java [hunk 1]
```diff
@@ -83,12 +83,17 @@ public QueryResult run(final WorkManager workManager, final WebUserConnection we
...
-    logger.debug("Wait until the query execution is complete or there is error submitting the query");
+    if (logger.isDebugEnabled()) {
+      logger.debug("Wait until the query execution is complete or there is error submitting the query");
+    }
     do {
       try {
-        isComplete = webUserConnection.await(TimeUnit.SECONDS.toMillis(1)); /*periodically timeout to check heap*/
-      } catch (Exception e) { }
-
+        isComplete = webUserConnection.timedWait(TimeUnit.SECONDS.toMillis(1)); /*periodically timeout 1sec to check heap */
```
Comment (sohami) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java]: Actually you don't need this `timedWait` method. You can still use same `await(timeout)` as you were using earlier. But outside while loop you can do below check to get the exception in error condition. Check [here](https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/work/prepare/PreparedStatementProvider.java#L158) 

```
// check if query completed with exception
if (isComplete && webUserConnection.getError() != null) {
      throw new UserRemoteException(webUserConnection.getError());
}
```

Comment (kkhatua) [reply exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java]: For some strange reason, the webUserConnection.getError() was not giving me the Exception message for invalid schemas. Let me see if I missed something.

Comment (kkhatua) [reply exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java]: Yep. It's sufficient to just wrap the DrillPBError into the UserException. I'm getting the invalid schema exception reported as well.

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java [hunk 2]
```diff
@@ -83,12 +84,13 @@ public QueryResult run(final WorkManager workManager, final WebUserConnection we
     float usagePercent = getHeapUsage();
 
     // Wait until the query execution is complete or there is error submitting the query
-    logger.debug("Wait until the query execution is complete or there is error submitting the query");
+    if (logger.isDebugEnabled()) {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java]: No need do this check, it's done internally.

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java [hunk 3]
```diff
@@ -83,12 +84,13 @@ public QueryResult run(final WorkManager workManager, final WebUserConnection we
...
+      logger.debug("Wait until the query execution is complete or there is error submitting the query");
+    }
     do {
       try {
-        isComplete = webUserConnection.await(TimeUnit.SECONDS.toMillis(1)); /*periodically timeout to check heap*/
-      } catch (Exception e) { }
-
+        isComplete = webUserConnection.await/*timedWait*/(TimeUnit.SECONDS.toMillis(1)); //periodically timeout 1sec to check heap
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java]: `/*timedWait*/` -> please remove this is obvious
`1sec` -> space

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java [hunk 4]
```diff
@@ -97,23 +99,29 @@ public QueryResult run(final WorkManager workManager, final WebUserConnection we
 
     //Fail if nearly out of heap space
     if (nearlyOutOfHeapSpace) {
+      UserException almostOutOfHeapException = UserException.resourceError(
+          new Throwable(
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java]: I know it was here before, but still why Throwable?

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java [hunk 5]
```diff
@@ -97,23 +99,29 @@ public QueryResult run(final WorkManager workManager, final WebUserConnection we
...
-                    )
-                )
-              .build(logger)
-            );
+        .addToEventQueue(QueryState.FAILED, almostOutOfHeapException);
+      //Return NearlyOutOfHeap exception
+      throw almostOutOfHeapException;
     }
 
     if (logger.isTraceEnabled()) {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java]: I know it was before but this check is not necessary.

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java (hunk 3 above)
Comment (kkhatua) [reply exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java]: My bad. That was actually a method I introduced before and reverted in the last commit.

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java (hunk 2 above)
Comment (kkhatua) [reply exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java]: Ok.

**Comment by kkhatua** [comment]
Done with the change

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java [hunk 6]
```diff
@@ -97,21 +97,22 @@ public QueryResult run(final WorkManager workManager, final WebUserConnection we
...
-                    "There is not enough heap memory to run this query using the web interface. "
-                    + "Please try a query with fewer columns or with a filter or limit condition to limit the data returned. "
-                    + "You can also try an ODBC/JDBC client. "
-                    )
-                )
-              .build(logger)
-            );
+        .addToEventQueue(QueryState.FAILED, almostOutOfHeapException);
+      //Return NearlyOutOfHeap exception
+      throw almostOutOfHeapException;
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java]: We did not throw exception before, why we are throwing it now?

Comment (kkhatua) [reply exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java]: I added the exception originally to the event queue, but I'm not sure if that will necessarily propagate the exception back. If it isn't thrown back, there is a possibility that before the eventQueue is handled, the resultSet will make it back to the WebServer  that will start constructing the JSONResponse object and run out of memory there.

**Pull request merged**